src/input.rs

```diff
@@ -13,40 +13,141 @@ use log::debug;
 use parking_lot::{RwLock, RwLockUpgradableReadGuard};
 use rustc_hash::FxHashMap;
 use std::collections::hash_map::Entry;
+use std::marker::PhantomData;
 
 /// Input queries store the result plus a list of the other queries
 /// that they invoked. This means we can avoid recomputing them when
 /// none of those inputs have changed.
-pub struct InputStorage<DB, Q>
+pub struct InputStorage<DB, Q, IP>
 where
     Q: Query<DB>,
     DB: Database,
-    Q::Value: Default,
+    IP: InputPolicy<DB, Q>,
 {
     map: RwLock<FxHashMap<Q::Key, StampedValue<Q::Value>>>,
+    input_policy: PhantomData<IP>,
+}
+
+pub trait InputPolicy<DB, Q>
```

Comment:
Note that we have

Memoization**Policy**
Sweep**Strategy**
Input**Policy**

Should we rename strategy?

Reply:
Probably, yes.

Reply:
They seem different to me but not for any good reason I guess.

Reply:
Seems like it should go in another PR though